### PR TITLE
SCI: move specific graphics code for PC-98 and Korean fan-patched versions into gfx driver classes

### PR DESCRIPTION
**Pull request by athrxx**
In one of the last gfx PRs I mentioned the idea of moving all the upscaling code into drivers classes to further clean up the engine code.

I now tried this experiment for the PC-98 versions. Fortunately, the original interpreters and drivers handle it in the same way, more or less (the graphics drivers' "normal" blitting opcode will upscale everything, only the font rendering is handled separately, for SCI0 the interpreters do it directly, for SCI1 there is a separate driver opcode). 

As a bonus, there are new 8-colors video modes for PC-98 and I also fixed some details in the 16-colors mode (the palette colors, font rendering, mouse cursor, text color).

At last, I also attached the Korean versions to a basic upscaled driver, since the goal still is to get rid of that code in the engine. Currently, there is no EGA or VGA/BW mode for Korean, since I only have the fan patch for SQ4 DOS/CD. Unfortunately, these fan patches are very difficult to find.

If you guys think that this is the way to go, I can try to wrap up the Mac games, too (which would be the final step for removing the upscaling code from the engine). There is already some skeleton code for Mac, but it is currently useless/disabled.

### 16 Colors:
Screenshot 2024-08-02 005137
### 8 Colors:
Screenshot 2024-08-02 005234
### 16 Colors:
Screenshot 2024-08-03 211647
### 8 Colors:
Screenshot 2024-08-03 211457
### 16 Colors:
Screenshot 2024-08-03 233039
### 8 Colors:
Screenshot 2024-08-03 232725
### 16 Colors:
Screenshot 2024-08-03 233245
### 8 Colors:
Screenshot 2024-08-03 232935
### 16 Colors:
Screenshot 2024-08-02 230952
### 8 Colors:
Screenshot 2024-08-02 230910

### CODE REVIEW

**Review by bluegr — APPROVED** [review]
As usual, great work! It's great to have the code for this functionality cleanly separated.
Also, thanks for clearing up the PC-98 color modes.

I can't say that the output of the 8 color mode is pretty, but it's nice for a bit of nostalgia :)

**Comment by athrxx** [comment]
> As usual, great work! It's great to have the code for this functionality cleanly separated. Also, thanks for clearing up the PC-98 color modes.
> 
> I can't say that the output of the 8 color mode is pretty, but it's nice for a bit of nostalgia :)

Thanks! I kind of like the (more elaborate) 8 colors mode for SCI1. But I even played through QGF1 in Hercules. I guess, if you're the one to implement these modes, you become more tolerant... ;-)

I noticed that we have a glitch with the Japanese line breaks.
My guess would be that there is a line break symbol that does a double paragraph. But for the person who implemented the Japanese text parsing it is probably a quick fix...

### Original:
Screenshot 2024-08-03 191550

### ScummVM:
### (without the changes of the PC, so please ignore the text color/style, **this is only about the missing empty lines**)
Screenshot 2024-08-03 191203

**Comment by athrxx** [comment]
I have added a fix for bug https://bugs.scummvm.org/ticket/15309, since it is not an urgent fix and otherwise I would have to rebase it on master and then probably force-push this branch.

The fix concerns the mysterious SCI_VERSION_01 (which has to be treated more like SCI1 in this case. For PC-98, this had already been addressed in this PR, now it is just extended to all targets that need it.

Of course, the fix can be based on current master if the PR gets rejected.

File: engines/sci/graphics/screen.cpp
Comment (sluicebox) [inline engines/sci/graphics/screen.cpp:206]: Did you mean to include this?

File: engines/sci/graphics/gfxdrivers.h
Comment (sluicebox) [inline engines/sci/graphics/gfxdrivers.h:219]: Did you mean to include this?

File: common/rendermode.h
Comment (sluicebox) [inline common/rendermode.h:57]: Unit tests need to be updated with this; this is currently breaking the github builds. See `test/common/rendermode.h`

**Review by sluicebox — CHANGES_REQUESTED** [review]

File: engines/sci/graphics/screen.cpp
Comment (athrxx) [reply engines/sci/graphics/screen.cpp:206]: It is kind of what I said in the description about "some skeleton code for Mac, but it is currently useless/disabled". So the answer to your question would be yes. But I can throw it out for sure. I was just wary of messing up my local branches too much. But removing that would actually relieve me of any pressure to deliver anything in the near future, so probably a good idea. I'll add a cleanup commit...

File: common/rendermode.h
Comment (athrxx) [reply common/rendermode.h:57]: Okay, I have too look into this. I have absolutely no clue what that is about. Probably something new...

**Comment by sluicebox** [comment]
Amazing. As with your other graphics driver changes, I'm thrilled with this improved code structure. The graphics code you're improving, especially screen initialization, has been notably difficult for me to navigate.

You're the graphics expert, so I just want to come up with things to test. Off the top of my head: 

- Native Macintosh fonts
- GK1 Macintosh (hi-res)
- PQ2 Japanese escaped newlines (I fixed these in 2021: 8a87d1030c706db7fede0b5ece0aeabdf1a671cb)
- PQ2 Japanese unescaped newlines
- KQ6 hi-res
- KQ6 Korean
- GK1 Korean (and I think I've got some other Korean fan patches laying around)

I'm very interested in the PQ2 newline bug you found. Did that occur before your graphics drivers? There was effectively no support for these until 2021, so if I missed something back then, I want to fully understand that first.

I'm in favor of merging this once I've run it through all the tests I'm dreaming up; probably at the end of this week. I want to do heavy up-front testing on this because it affects relatively obscure games. In my experience, it takes a long time to hear about problems in these versions, if we ever do. Last month my tools discovered that Macintosh Castle of Dr. Brain has *never* been completable in ScummVM; it's always crashed right away in the same scene: 157c8935f30d763f11d9ee20a3d16a5aa477cdd5 .

**Comment by athrxx** [comment]
Thanks!

Testing this for a while is a good idea for sure. I usually do test KQ6 Mac, since it has that interesting mode for the video playback.

KQ6 Mac has some mouse cursor and icon bar glitches in rgb rendering mode for which I do have local fixes. At least the one for the icon bar could probably be pushed without any overhead (=the commented out, now removed code).

Also, I noticed that the engine uses OSystem::warpMouse() which won't work with the scaled gfx drivers like that. I also have a local fix for that (it goes on top of this PR, so it doesn't currently make sense to make another PR for it).

The PQ2 newline bug is not something caused by me (otherwise I wouldn't present it like that ;-) ). The screenshot is from before. You can tell from the black text color and non-fat text drawing. With this PR the text will be blue and in fat print, but it doesn't do a thing about these line breaks. Unfortunately, a quick debug run didn't give me any clues. The lines all seem to have a `\n` ending. But it is a good thing that you're actually the one who did that coding, so we might find something eventually...

If you do have the KQ6 Korean patch, could you please send me that? I am sure the Koreans can hardly wait to finally get their EGA mode support.

File: common/rendermode.h
Comment (athrxx) [reply common/rendermode.h:57]: It isn't actually new. It can be seen from the things that have been included. All render modes that came later than kRenderPC9821 and kRenderPC9801 aren't in...

**Comment by sluicebox** [comment]
Quick test results:

- PQ2 Japanese mouse cursor is squished for me. Default game settings. Don't think I have any weird global video settings beyond SDL instead of OpenGL. (Windows)
- PQ2 Japanese newline bug predates my 2021 fixes, so we can ignore it on this PR.
- GK1 Mac was broken (by me!) so I just fixed that. Another obscure version that's been busted for a while. Hi-res text looks good.
- FPFP Mac looks good (native Mac fonts, so upscaled)
- KQ6 Mac looks good (same)
- SQ4CD Korean looks good
- KQ6 CD w/ hi-res portraits looks good

I only still have SQ4CD and FPFP (CD?) Korean, and FPFP never ended up in the detection tables. I'll keep looking for the others.

scummvm-pq2-pc98-ja-00000

**Review by sluicebox — APPROVED** [review]

**Comment by bluegr** [comment]
Thanks @athrxx for your amazing work, and thanks @sluicebox for testing :)

I don't have anything else to add here, this is tested and good to be merged now, so I'm hitting the merge button.

**Comment by sluicebox** [comment]
i did not approve this, i wasted my time writing many words explaining i wanted to test it for a week

**Comment by athrxx** [comment]
> * PQ2 Japanese mouse cursor is squished for me. Default game settings. Don't think I have any weird global video settings beyond SDL instead of OpenGL. (Windows)

Heya, by "squished" you mean the vertical only scale, right? I did this on purpose, because the original doors it. You can even see it in the emulator screen shot I made above to demonstrate the line break bug.

This is basically what the original PC-98 gfx drivers do: 
- QFG does not scale the cursor at all
- PQ2 does the vertically only scale
- SCI1 scales both vertically and horizontally

I replicated all of that, but I understand that the aesthetics are debatable. It is a simple bool setting in the drivers for the h and v setting. So we can easily have that as we want...

**Comment by sluicebox** [comment]
Just noticed that this caused some Korean text to no longer be rendered in SQ4CD: https://bugs.scummvm.org/ticket/15375

Oh *now* I test it... =)

**Pull request merged**
